### PR TITLE
Feature/species entry validation

### PR DESCRIPTION
**Pull request by TurtleCubed**

### CODE REVIEW

File: BirdQuiz/species_validation.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import requests
+import pandas as pd
+from pandas import DataFrame
```
Comment (aaronsun1030) [inline BirdQuiz/species_validation.py]: pandas seems like overkill just to read a csv, why not use built-in `csv` library instead? Will pandas be used in other places?

File: BirdQuiz/species_validation.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from fast_autocomplete import AutoComplete
```
Comment (aaronsun1030) [inline BirdQuiz/species_validation.py]: Not in `requirements.txt`, nor is it ever used. Also there's already an autocomplete box functionality `AutocompleteEntryListbox` which I would check out.

File: BirdQuiz/QuizFrontend.py [hunk 3]
```diff
@@ -31,7 +33,7 @@ def show_intro(self):
         self.root.geometry('1280x720')
         self.label.configure(text = "Welcome to BirdQuiz! Press play to begin:")
 
-        im = self.resize_to_tk(Image.open("./resources/intro.jpg"))
+        im = self.resize_to_tk(Image.open("./BirdQuiz/resources/intro.jpg"))
```
Comment (aaronsun1030) [inline BirdQuiz/QuizFrontend.py]: I think we should assume that the code is run in the `BirdQuiz` directory.

File: BirdQuiz/species_validation.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import os
+
+CHECKLIST_URL = "https://com-aab-media.s3.amazonaws.com/common/ebird_taxonomy_v2022.csv"
+CHECKLIST_NAME = "ebird_taxonomy_v2022.csv"
+
+class Validate():
+
+    def __init__(self) -> None:
+        """Gets the 2022 taxononmy, and initalizes the validation set"""
+        if not os.path.isfile("BirdQuiz/resources/" + CHECKLIST_NAME):
```
Comment (aaronsun1030) [inline BirdQuiz/species_validation.py]: (repetitive) I think we should assume that the code is run in the `BirdQuiz` directory.

File: BirdQuiz/species_validation.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            self.get_checklist_from_url()
+        df = pd.read_csv("BirdQuiz/resources/" + CHECKLIST_NAME)
+        self.species_set = set(df["PRIMARY_COM_NAME"])
+
+    def get_checklist_from_url(self):
+        """Downloads the 2022 eBird taxonomy list to resources"""
+        with open(CHECKLIST_NAME, "wb") as f:
```
Comment (aaronsun1030) [inline BirdQuiz/species_validation.py]: Saves to base directory, not `resources` directory.

File: BirdQuiz/QuizFrontend.py [hunk 6]
```diff
@@ -50,8 +52,8 @@ def show_intro(self):
         # Buttons to add/remove species
         addbutton = Button(self.root, width=1, height=1)
         addbutton.grid(row=1, column=5)
-        addbutton.configure(text = '+', bd = '4', command = lambda: [listbox.insert(0, inputbox.get("1.0", "end")), inputbox.delete("1.0", "end")])
-        self.root.bind('<Return>', lambda _: [listbox.insert(0, inputbox.get("1.0", "end")), inputbox.delete("1.0", "end")])
+        addbutton.configure(text = '+', bd = '4', command=lambda i=inputbox, l=listbox: self.add_to_listbox(i, l))
+        self.root.bind('<Return>', lambda i=inputbox, l=listbox: self.add_to_listbox(i, l))
```
Comment (aaronsun1030) [inline BirdQuiz/QuizFrontend.py]: Errors on return, need an extra argument in lambda function

**Review by aaronsun1030 — CHANGES_REQUESTED** [review]

**Comment by aaronsun1030** [comment]
Weird bug with `.title()` not handling apostrophes well:

Capitalizes after apostrophe (baird's sandpiper -> Baird'S Sandpiper), see [stack overflow](https://stackoverflow.com/questions/8199966/python-title-with-apostrophes).

After inputting this, nothing works anymore. Also, bizarre ghost printing:

```
S is an invalid species name
S
Baird'S Sandpiper is an invalid species name
S
Baird'S Sandpiper
Mallard is an invalid species name
S
Baird'S Sandpiper
Mallard
Baird'S Sandpiper is an invalid species name
```

File: BirdQuiz/species_validation.py (hunk 1 above)
Comment (TurtleCubed) [reply BirdQuiz/species_validation.py]: pandas will be used in other places, should I leave it out of this branch and only add it when necessary?

File: BirdQuiz/species_validation.py (hunk 2 above)
Comment (TurtleCubed) [reply BirdQuiz/species_validation.py]: My bad on this one. I was testing out fast_autocomplete for a different feature.

File: BirdQuiz/QuizFrontend.py (hunk 3 above)
Comment (TurtleCubed) [reply BirdQuiz/QuizFrontend.py]: kk, will change

File: BirdQuiz/species_validation.py (hunk 4 above)
Comment (TurtleCubed) [reply BirdQuiz/species_validation.py]: done

File: BirdQuiz/species_validation.py (hunk 5 above)
Comment (TurtleCubed) [reply BirdQuiz/species_validation.py]: yep got that in another commit

File: BirdQuiz/QuizFrontend.py (hunk 6 above)
Comment (TurtleCubed) [reply BirdQuiz/QuizFrontend.py]: that's really odd, but I got that too

**Comment by TurtleCubed** [comment]
> Weird bug with `.title()` not handling apostrophes well:
> 
> Capitalizes after apostrophe (baird's sandpiper -> Baird'S Sandpiper), see [stack overflow](https://stackoverflow.com/questions/8199966/python-title-with-apostrophes).
> 
> After inputting this, nothing works anymore. Also, bizarre ghost printing:
> 
> ```
> S is an invalid species name
> S
> Baird'S Sandpiper is an invalid species name
> S
> Baird'S Sandpiper
> Mallard is an invalid species name
> S
> Baird'S Sandpiper
> Mallard
> Baird'S Sandpiper is an invalid species name
> ```

Regarding this: The issue was that upon failing validation, the inputbox would not clear. The ghost printing was from the inputbox having an extra line containing the previous incorrect string.

File: BirdQuiz/species_validation.py (hunk 1 above)
Comment (aaronsun1030) [reply BirdQuiz/species_validation.py]: no IDC

File: BirdQuiz/QuizFrontend.py [hunk 7]
```diff
@@ -172,7 +174,14 @@ def end(self):
         self.root.mainloop()
 
     
-
+    def add_to_listbox(self, inputbox, listbox):
+        """Moves text from the inputbox to the listbox with some text validation"""
+        text = inputbox.get("1.0", "end").strip()
```
Comment (aaronsun1030) [inline BirdQuiz/QuizFrontend.py:179]: I like the auto-capitalization, so can you add `capwords` like they recommend in [this post](https://stackoverflow.com/questions/8199966/python-title-with-apostrophes)?

Comment (aaronsun1030) [reply BirdQuiz/QuizFrontend.py:179]: to be fixed in next PR

**Pull request merged**